[PR] Edit `timer`

LuchoTurtle opened the pull request:
Should close #195 

Still a work in progress. The user when editing an item will see a list of timers that he can edit.

LuchoTurtle commented:
Even though this feature should be and is super simple to implement, I'm getting the go around implementing proper form validation and getting updated structs from the LiveView to the Phoenix server so I can update timers with their IDs.

This shouldn't be hard but it's weird because the way `form` is implemented when editing an item should also work on this occasion but I'm getting 0 events in the Phoenix server 😞 . 

I'll keep trying

nelsonic commented:
@LuchoTurtle thanks for opening this PR to track/share your progress! 🎉
Don’t worry about something not working _first_ time. That’s normal. Just keep sharing your progress/frustrations and if you hit a brick wall, give as much detail as possible so one of us can help you figure it out. 👌

nelsonic commented on lib/app_web/live/app_live.ex:
```diff
@@ -124,17 +124,16 @@ defmodule AppWeb.AppLive do
-        start: start,
-        stop: stop
-      })
+      case DateTime.compare(start, stop) do
```
nice. 👌

LuchoTurtle commented:
Slow progress so far, from the user story perspective. Since I was practically _obliged_ to use [Bluzkly's datetime parser gist](https://gist.github.com/bluzky/62a20cdb57b17f47c67261c10aa3da8b) (even more so if we are in the future wanting to accommodate other formats), since I was stuck with sending error feedback from the server to the form, I tried to focus on testing what I had done.

Since I have to test this parser, I had to battle a bit with VSCode to _try_ and get debugging working but I never did (got this `The database for App.Repo couldn't be dropped: ERROR 55006 (object_in_use) database "app_test" is being accessed by other users` error), so I've just been `Logger.debug`ing my way through. It's taking a lot more than expected, it's hard to get code coverage on edge cases...

nelsonic commented:
Please don't worry about things taking longer than you expected. ⏳ 
Just stay focussed on learning, capturing your learning 📝 
and the feature will be a _byproduct_ of that learning. 🪄 
Very few companies/orgs think this way; we do! 👌 
remember `Logger.debug` --> [`dbg/1`](https://elixir-lang.org/getting-started/debugging.html#dbg)

If you want to remote pair on anything please LMK (I'm _always_ available!) 👨‍💻 
But equally I understand the merits of "figuring it out for yourself" ... 😉

LuchoTurtle commented:
Added all tests so coverage should all be sorted. Form validation upon submission should now work properly, with feedback to the user. 

Experimented with a few options about `schemaless changesets`, `nested form changesets` but those weren't working properly. So I went with the route of adding a list of `editing_timers` (which is an array of `Timer` changesets) that are assigned to the socket dynamically whenever a user edits an item. This is better than the initial version that fetched **all nested timers** on `mount`, which could cause slow mount times in the future with a large amount of `timers`. 

An `index` input field was also added. This is only for optimization purposes. If we know the index of the timer we want to edit within the changeset timer array, we can update the array with `O(1)` by just replacing the element with index access.

LuchoTurtle commented:
Added some docs and changed the BUILD file with updated information 😄 .

nelsonic commented on .gitignore:
```diff
@@ -46,4 +46,5 @@ npm-debug.log
 #elm
-elm-stuff/
\ No newline at end of file
+elm-stuff/
```
What `elm-stuff` is happening in the project? 😄

LuchoTurtle replied:
No idea how that got into the project, I recall adding `.vscode` because of my launch settings. Never done Elm in my life 😅

nelsonic commented:
Ran:
```sh
git checkout edit-timer
mix deps.get
mix c
```

Got:

```sh
Compiling 29 files (.ex)
warning: variable "e" is unused (if the variable is not meant to be used, prefix it with an underscore)
  lib/app_web/live/app_live.ex:214: AppWeb.AppLive.handle_event/3

warning: clauses with the same name and arity (number of arguments) should be grouped together, "def handle_event/3" was previously defined (lib/app_web/live/app_live.ex:33)
  lib/app_web/live/app_live.ex:164

warning: module attribute @nodoc was set but never used
  lib/app_web/live/app_live.ex:132

Generated app app

08:44:18.521 [info] extension "citext" already exists, skipping
warning: function create_profile_with_no_name/1 is unused
  test/app_web/controllers/profile_controller_test.exs:75

.warning: variable "person" is unused (if the variable is not meant to be used, prefix it with an underscore)
  test/app_web/controllers/tag_controller_test.exs:97: AppWeb.TagControllerTest.create_person/1

warning: variable "now" is unused (if the variable is not meant to be used, prefix it with an underscore)
  test/app_web/live/app_live_test.exs:140: AppWeb.AppLiveTest."test update an item's timer"/1
```

I can address these as part of my review or you can tidy up the warnings and re-assign to me. 💭

LuchoTurtle commented:
I got it, thanks for the feedback 👍

LuchoTurtle commented:
Should be fixed, sorry about that.

nelsonic commented:
@LuchoTurtle nice one. Thanks for updating. I will continue reviewing after standup. :ok_hand:

nelsonic commented on BUILDIT.md:
```diff
@@ -2756,6 +2756,804 @@ Learn more about Ecto with the guides documention, especially the How to section
 https://hexdocs.pm/ecto/getting-started.html (taken from: https://dashbit.co/ebooks/the-little-ecto-cookbook)
 
 
```
Table of contents in `BUILDIT.md` not updated to include Edit Timers:
https://github.com/dwyl/mvp/blob/d601dc0863feef976285257883a815acca323258/BUILDIT.md

<img width="1114" alt="image" src="https://user-images.githubusercontent.com/194400/200891409-0bfa8cff-ec22-46c3-b896-e0986c8bbf4a.png">

Definitely don't do this _manually_, use: https://marketplace.visualstudio.com/items?itemName=yzhang.markdown-all-in-one

nelsonic replied:
Fixed by: https://github.com/dwyl/mvp/pull/196/commits/0f0b8139e67e91a77c1f5cee8bf32e04a71dbe22

LuchoTurtle commented:
Just pushed an overlap validation on the timer being updated with the others of the same item. I will get on with having the `stop` field _not required_ and being possible to edit 👌

LuchoTurtle commented:
@nelsonic I've added the feature and the referring tests you asked for. However, I won't ask you to review it just yet, I want to make it prettier (as you mentioned at the standup) and make it possible to delete timers. 👌

nelsonic commented:
@LuchoTurtle tidying the UI/UX can be done in a subsequent iteration. ✨ 
And **_`deleting`_ `timers`** isn't actually something we _want_ at this point. 🔪 
It could certainly be a feature in the future. But right now none of our "Alpha" testers have requested it. 💭

LuchoTurtle commented:
@nelsonic fair enough. Though I think deleting should be a part of the whole "editing timers" panorama. Sometimes it's easier to delete a timer when editing one, making it is much faster to manage overlapping scenarios.

It's your call though 🎉

nelsonic commented:
@LuchoTurtle please open an issue to describe the use-case/need for deleting `timers`. 🆕 
We've not had the feature/requirement described anywhere (_yet_) so I'm curious what the real-world scenario would be. 💭 
Going to take another look at the PR now. 👌

nelsonic commented on BUILDIT.md:
```diff
@@ -2756,11 +2766,809 @@ Learn more about Ecto with the guides documention, especially the How to section
+  def build_regex(format) do
+    keys = Map.keys(@mapping) |> Enum.join("")
+
+    Regex.compile!("([^%]*)%([#{keys}])([^%]*)")
```
This `Regex` should be clarified, ideally with examples. 💭

nelsonic commented on lib/app/datetime_parser.ex:
```diff
@@ -0,0 +1,196 @@
+# Full credit of this module goes to https://dev.to/onpointvn/build-your-own-date-time-parser-in-elixir-50be
+# Do check the gist -> https://gist.github.com/bluzky/62a20cdb57b17f47c67261c10aa3da8b
+defmodule App.DateTimeParser do
```
@LuchoTurtle one of the reasons I requested that you leave regular comments on the issues you're working on is precisely to know in advance when/why we _need_ to include a Date Parsing library within the MVP. 
This is **_library_ code** not **business logic**. 💭 

The actual **_edit_ `timer`** functionality should be a comparatively _tiny_ amount of code. 
Which Date/Time Parsing libraries did you look at _before_ deciding to write this File? ❓ 

We've used **`Timex`** in the past: 
https://hexdocs.pm/timex/basic-usage.html#formatting-datetimes

Which function did you need that you didn't find in one of the existing libraries? 💭

LuchoTurtle replied:
[I did leave a comment regarding my decision of using a library code.](https://github.com/dwyl/mvp/pull/196#issuecomment-1297095672). Although I wasn't sure where I should put it in the code at first (I agree, it's not business logic, it should go to an `utils` folder or some sort), I decided to not use `Timex` because I didn't think it was worth it to import a whole library just for the sole purpose of parsing a date on a specified format. I tried to keep things lightweight on this one.

But if you think using `Timex` makes everything cleaner, I can change, no prob 😄

nelsonic commented on lib/app_web/live/app_live.ex:
```diff
@@ -108,15 +116,246 @@ defmodule AppWeb.AppLive do
+    index = String.to_integer(index)
+    changeset_obj = Enum.at(timer_changeset_list, index)
+
+    try do
```
This is quite a lot of business logic to include in `app_live.ex` ...
Is there a reason we can't put most of the `timer` related code in `/lib/app/timer.ex` ?

LuchoTurtle replied:
You're right, it should go into the `timer.ex` file, as it pertains to it. I was going to refactor the code and hopefully make the flow easier to understand before pushing to review 😃

nelsonic commented on lib/app_web/live/app_live.ex:
```diff
@@ -108,15 +116,246 @@ defmodule AppWeb.AppLive do
+  # - `changeset_error_key`: atom key of the changeset object you want to associate the error message
+  # - `changeset_error_message`: the string message to error the changeset key with.
+  @doc false
+  defp error_timer_changeset(
```
This `timer_changeset` error handling code should _definitely_ not be in the `lib/app_web/live/app_live.ex` file.

LuchoTurtle commented:
@nelsonic I don't know why you're reviewing the code right now. Before the discussion about deleting timers, I was going to refactor the code before submitting it for review (hence why I still had the "in-progress" label :P), since I know a chunk of the business logic was in the `app_live.ex` file.
I thank you for the feedback nonetheless. 🎉

nelsonic commented:
Ah, apologies. I didn't expect anyone to be online at 22:00 on a Saturday evening. 💭 
Assigning back to you. 👤 
Will get some sleep instead. 👌

LuchoTurtle commented:
@nelsonic just updated the refactored code, with the addition that now the user knows whether it `start` or `stop` is invalid. Error handling should be much more readable and the business logic should be well-placed.
Tests and `build.md` file are both also updated :)

nelsonic commented on lib/app/timer.ex:
```diff
@@ -1,15 +1,17 @@
 defmodule App.Timer do
   use Ecto.Schema
   import Ecto.Changeset
+  alias App.Item
```
Does aliasing `App.Item` in the `timer.ex` 
and aliasing `App.Timer` in `item.ex`
create a circular reference? 
(I don't _think_ it does, but wondering ...)

LuchoTurtle replied:
Hm, I don't believe using `alias` would create a circular reference (that's probably [why it's recommended instead of using `import`](https://stackoverflow.com/questions/70107659/in-elixir-why-is-alias-preferred-over-import-for-importing-the-modules). Regardless, this shouldn't be a problem or else everytime we wanted to reference another object in an Ecto schema as *belongs_to* or something similar, we'd have this problem 🤔

nelsonic commented on lib/app/timer.ex:
```diff
@@ -63,6 +65,278 @@ defmodule App.Timer do
+  Otherwise, it returns {:error, updated_list}, where `error_term` is the error that occurred and `updated_list` being the updated item changeset list with the error.
+  """
+  def update_timer_inside_changeset_list(
+        timer_id,
```
could the function signature here be a Map/Struct with these fields? 💭

LuchoTurtle replied:
Just pushed a commit with this in mind. I didn't put *everything* into a struct but all the timer-related ones. The index and changeset list stayed the same 😄

nelsonic commented on mix.exs:
```diff
@@ -54,6 +54,7 @@ defmodule App.MixProject do
       {:telemetry_poller, "~> 1.0"},
       {:jason, "~> 1.2"},
       {:plug_cowboy, "~> 2.5"},
+      {:timex, "~> 3.7"},
```
```suggestion
      # Time string parsing: github.com/bitwalker/timex
      {:timex, "~> 3.7"},
```

nelsonic commented on test/app/item_test.exs:
```diff
@@ -141,7 +141,12 @@ defmodule App.ItemTest do
+        Timer.start(%{
+          item_id: item1.id,
+          person_id: 1,
+          start: started,
```
Should having the `timer.start` and `timer.start` the same return an `:error_start_equal_stop` instead of `{:ok, timer}` now? or am I reading the code incorrectly?

LuchoTurtle commented:
Just committed the changes requested 😄

nelsonic commented:
When attempting to edit a timer, getting the following error in console:

```sh
[error] GenServer #PID<0.728.0> terminating
** (FunctionClauseError) no function clause matching in NaiveDateTime.compare/2
    (elixir 1.14.1) lib/calendar/naive_datetime.ex:1025: NaiveDateTime.compare(~N[2022-11-16 15:40:36], nil)
    (app 1.0.0) lib/app/timer.ex:205: anonymous fn/4 in App.Timer.update_timer_inside_changeset_list/3
    (elixir 1.14.1) lib/enum.ex:2468: Enum."-reduce/3-lists^foldl/2-0-"/3
    (app 1.0.0) lib/app/timer.ex:199: App.Timer.update_timer_inside_changeset_list/3
    (app 1.0.0) lib/app_web/live/app_live.ex:143: AppWeb.AppLive.handle_event/3
    (phoenix_live_view 0.18.2) lib/phoenix_live_view/channel.ex:382: anonymous fn/3 in Phoenix.LiveView.Channel.view_handle_event/3
    (telemetry 1.1.0) /mvp/deps/telemetry/src/telemetry.erl:320: :telemetry.span/3
    (phoenix_live_view 0.18.2) lib/phoenix_live_view/channel.ex:216: Phoenix.LiveView.Channel.handle_info/2
    (stdlib 4.1.1) gen_server.erl:1123: :gen_server.try_dispatch/4
    (stdlib 4.1.1) gen_server.erl:1200: :gen_server.handle_msg/6
    (stdlib 4.1.1) proc_lib.erl:250: :proc_lib.wake_up/3
```

Can show you on Zoom if you have time. 🧑‍💻

LuchoTurtle commented:
Sure thing, I'm joining the standup link

LuchoTurtle commented:
Apparently, this bug happens when I'm editing a timer and try to edit an old timer.
Additionally, I ought to check updating the cumulative timer count after updating.

nelsonic commented:
Editing times is _mostly_ working for me but I occasionally get these errors:

```sh
[error] GenServer #PID<0.722.0> terminating
** (FunctionClauseError) no function clause matching in NaiveDateTime.compare/2
    (elixir 1.14.1) lib/calendar/naive_datetime.ex:1025: NaiveDateTime.compare(~N[2022-11-16 17:21:07], nil)
    (app 1.0.0) lib/app/timer.ex:205: anonymous fn/4 in App.Timer.update_timer_inside_changeset_list/3
    (elixir 1.14.1) lib/enum.ex:2468: Enum."-reduce/3-lists^foldl/2-0-"/3
    (app 1.0.0) lib/app/timer.ex:199: App.Timer.update_timer_inside_changeset_list/3
    (app 1.0.0) lib/app_web/live/app_live.ex:143: AppWeb.AppLive.handle_event/3
    (phoenix_live_view 0.18.2) lib/phoenix_live_view/channel.ex:382: anonymous fn/3 in Phoenix.LiveView.Channel.view_handle_event/3
    (telemetry 1.1.0) /mvp/deps/telemetry/src/telemetry.erl:320: :telemetry.span/3
    (phoenix_live_view 0.18.2) lib/phoenix_live_view/channel.ex:216: Phoenix.LiveView.Channel.handle_info/2
    (stdlib 4.1.1) gen_server.erl:1123: :gen_server.try_dispatch/4
    (stdlib 4.1.1) gen_server.erl:1200: :gen_server.handle_msg/6
    (stdlib 4.1.1) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
```

The UI does not display an error and everything _appears_ to continue working. 
So from the `person` _using_ the App's perspective it appears to work.  👌 
Just want to make sure this error condition doesn't get committed to `main` and then we end up with tech-debt to pay. 💭

nelsonic commented:
Thanks for taking a look at this. 👌

LuchoTurtle commented:
I pushed the tests to 100% coverage @nelsonic . It seems [this commit](https://github.com/dwyl/mvp/pull/196/commits/38b26ae73b502d39083c71a7ee41d30f3f566b0b) broke some auth tests :/

nelsonic reviewed (approved):
Thanks for making these update @LuchoTurtle 👌 
Edit Timer MVP Looking good. Might require some UI/UX iteration. 💭 
But let's get some feedback from @iteles & @Stephanymtr ASAP. :shipit:

nelsonic commented:
Working in **`prod`**: https://mvp.fly.dev 🚀 

<img width="1056" alt="image" src="https://user-images.githubusercontent.com/194400/203267607-b2f99fdf-ca2c-4b5c-a82c-6207a91d7349.png">

Thanks again @LuchoTurtle 🙌
